### PR TITLE
[Feature] FindEdge/FindEdges for Immutable Graph

### PR DESCRIPTION
**Pull request by yzh119**
## Description
FindEdge(s) and EdgeSubgraph have not been implemented in Immutable Graph. #369 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- Add an `EdgeList` structure in `ImmutableGraph`, and a shared pointer attribute `edge_list_`; to avoid unnecessary memory cost, `edge_list_` is created when it's required.
- Complete `EdgeSubgraph`
  - [x] Still lacks test coverage for `EdgeSubgraph`, but I was too lazy and I would finish it soon.

### CODE REVIEW

**Comment by yzh119** [comment]
Works fine with Transformer.

**Comment by yzh119** [comment]
@zheng-da EdgeSubgraph finished.

File: src/graph/immutable_graph.cc [hunk 1]
```diff
@@ -158,14 +158,32 @@ class HashTableChecker {
   }
 };
 
+ImmutableGraph::EdgeList::Ptr ImmutableGraph::EdgeList::FromCSR(
+    std::vector<int64_t> *indptr, std::vector<dgl_id_t> *indices, std::vector<dgl_id_t> *edge_ids,
+    int sort_on) {
```
Comment (zheng-da) [inline src/graph/immutable_graph.cc]: `sort_on` is a little misleading. It indicates what `indptr` means instead of which endpoint is sorted.

File: src/graph/immutable_graph.cc [hunk 2]
```diff
@@ -197,6 +215,44 @@ std::pair<ImmutableGraph::CSR::Ptr, IdArray> ImmutableGraph::CSR::VertexSubgraph
...
+  }
+
+  const size_t n = oldv2newv.size();
+
+  std::vector<Edge> edges;
+  for (int64_t i = 0; i < len; i++) {
+    dgl_id_t src_id = oldv2newv.find(edge_list->src_points[eid_data[i]])->second;
+    dgl_id_t dst_id = oldv2newv.find(edge_list->dst_points[eid_data[i]])->second;
+    edges.push_back(Edge{src_id, dst_id, static_cast<dgl_id_t>(i)});
+  }
```
Comment (zheng-da) [inline src/graph/immutable_graph.cc]: why do you not merge this loop with the one above. we should avoid hashtable lookups as much as possible.

File: src/graph/immutable_graph.cc [hunk 3]
```diff
@@ -449,6 +505,36 @@ ImmutableGraph::EdgeArray ImmutableGraph::EdgeIds(IdArray src_ids, IdArray dst_i
...
+    src.push_back(edge.first);
+    dst.push_back(edge.second);
+  }
+
+  IdArray rst_src = IdArray::Empty({len}, eids->dtype, eids->ctx);
+  IdArray rst_dst = IdArray::Empty({len}, eids->dtype, eids->ctx);
+  dgl_id_t* rst_src_data = static_cast<dgl_id_t*>(rst_src->data);
+  dgl_id_t* rst_dst_data = static_cast<dgl_id_t*>(rst_dst->data);
+  std::copy(src.begin(), src.end(), rst_src_data);
+  std::copy(dst.begin(), dst.end(), rst_dst_data);
```
Comment (zheng-da) [inline src/graph/immutable_graph.cc]: why do you store data in the vector first and copy them? The length of the vectors is known.

File: tests/compute/test_graph.py [hunk 4]
```diff
@@ -184,10 +184,40 @@ def test_readonly():
...
+        assert fail
+
 if __name__ == '__main__':
     test_graph_creation()
     test_create_from_elist()
     test_adjmat_cache()
     test_incmat()
     test_incmat_cache()
     test_readonly()
+    test_find_edges()
```
Comment (zheng-da) [inline tests/compute/test_graph.py]: can you add a test on generating edge subgraphs?

File: include/dgl/immutable_graph.h [hunk 5]
```diff
@@ -32,6 +32,30 @@ class ImmutableGraph: public GraphInterface {
...
+      CHECK_LT(eid, src_points.size()) << "Invalid edge id " << eid;
+      src_points[eid] = src;
+      dst_points[eid] = dst;
+    }
+
+    static EdgeList::Ptr FromCSR(
+        std::vector<int64_t> *indptr,
+        std::vector<dgl_id_t> *indices,
+        std::vector<dgl_id_t> *edge_ids,
+        bool in_csr);
```
Comment (zheng-da) [inline include/dgl/immutable_graph.h]: can you make the input arguments const reference? input pointers usually mean the vectors will be modified by the method.

**Comment by zheng-da** [comment]
everything else looks good.

File: include/dgl/immutable_graph.h (hunk 5 above)
Comment (yzh119) [reply include/dgl/immutable_graph.h]: updated
